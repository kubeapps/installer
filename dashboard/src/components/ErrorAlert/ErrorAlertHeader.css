.error__icon {
  width: 30px;
  height: 30px;
  display: inline-block;
  border-radius: 50%;
  vertical-align: bottom;
  background: #c7c9c8;
}

.error__icon > svg {
  position: relative;
  color: #ffffff;
  top: 3px;
  left: 3px;
  margin: auto;
}
