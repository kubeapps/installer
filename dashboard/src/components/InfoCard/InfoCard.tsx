import * as React from "react";

import { Link } from "react-router-dom";
import Card, { CardContent, CardIcon } from "../Card";
import "./InfoCard.css";

export interface IServiceInstanceCardProps {
  title: string;
  info: string;
  link?: string;
  icon?: string;
  tag1Class?: string;
  tag1Content?: string;
  tag2Class?: string;
  tag2Content?: string;
}

const InfoCard: React.SFC<IServiceInstanceCardProps> = props => {
  const { title, link, icon, info, tag1Content, tag1Class, tag2Content, tag2Class } = props;
  return (
    <Card responsive={true} className="ListItem">
      <Link to={link || "#"} title={title}>
        <CardIcon icon={icon} />
        <CardContent>
          <div className="ListItem__content">
            <h3 className="ListItem__content__title type-big">{title}</h3>
            <div className="ListItem__content__info">
              <p className="margin-reset type-small padding-t-tiny type-color-light-blue">{info}</p>
              <div>
                {tag1Content && (
                  <span
                    className={`ListItem__content__info_tag ListItem__content__info_tag-1 type-small type-color-white padding-t-tiny padding-h-normal ${tag1Class ||
                      ""}`}
                  >
                    {tag1Content}
                  </span>
                )}
                {tag2Content && (
                  <span
                    className={`ListItem__content__info_tag ListItem__content__info_tag-2 type-small type-color-white padding-t-tiny padding-h-normal margin-b-normal ${tag2Class ||
                      ""}`}
                  >
                    {tag2Content}
                  </span>
                )}
              </div>
            </div>
          </div>
        </CardContent>
      </Link>
    </Card>
  );
};

export default InfoCard;
