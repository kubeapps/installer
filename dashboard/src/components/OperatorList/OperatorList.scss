.operatorListContainer {
  padding-bottom: 5vh;
}
