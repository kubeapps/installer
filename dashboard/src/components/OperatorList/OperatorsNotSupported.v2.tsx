import * as React from "react";
import { Link } from "react-router-dom";

import Alert from "components/js/Alert";
import * as url from "shared/url";

interface IOperatorNotSupportedProps {
  namespace: string;
  kubeappsCluster: string;
}

function OperatorNotSupported(props: IOperatorNotSupportedProps) {
  return (
    <Alert theme="warning">
      <h5>Operators are supported on the cluster on which Kubeapps is installed only</h5>
      <p>
        Kubeapps' Operator support enables the addition of{" "}
        <Link to={url.app.operators.list(props.kubeappsCluster, props.namespace)}>
          operators on the cluster on which Kubeapps is installed only
        </Link>
        .
      </p>
    </Alert>
  );
}

export default OperatorNotSupported;
