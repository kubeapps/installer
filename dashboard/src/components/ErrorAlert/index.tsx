import ErrorAlertHeader from "./ErrorAlertHeader";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export { ErrorAlertHeader, NotFoundErrorAlert, PermissionsErrorAlert, UnexpectedErrorAlert };
