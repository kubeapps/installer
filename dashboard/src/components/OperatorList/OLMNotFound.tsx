import * as React from "react";
import { Link } from "react-router-dom";

import { NotFoundErrorAlert } from "../ErrorAlert";

const OLMNotFound: React.SFC = () => {
  return (
    <NotFoundErrorAlert header="Operator Lifecycle Manager (OLM) not installed.">
      <div>
        <p>
          Ask an administrator to install the{" "}
          <a
            href="https://github.com/operator-framework/operator-lifecycle-manager"
            target="_blank"
          >
            Operator Lifecycle Manager
          </a>{" "}
          to browse, provision and manage Operators within Kubeapps.
        </p>
        <Link className="button button-primary button-small" to={"/charts/svc-cat/catalog"}>
          Install OLM
        </Link>
      </div>
    </NotFoundErrorAlert>
  );
};

export default OLMNotFound;
