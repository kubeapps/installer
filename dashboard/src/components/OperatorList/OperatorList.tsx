import { RouterAction } from "connected-react-router";
import * as React from "react";

import { ForbiddenError, IClusterServiceVersion, IPackageManifest } from "../../shared/types";
import { api, app } from "../../shared/url";
import { escapeRegExp } from "../../shared/utils";
import { CardGrid } from "../Card";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import InfoCard from "../InfoCard";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import OLMNotFound from "./OLMNotFound";

export interface IOperatorListProps {
  isFetching: boolean;
  checkOLMInstalled: () => Promise<boolean>;
  isOLMInstalled: boolean;
  namespace: string;
  getOperators: (namespace: string) => Promise<void>;
  operators: IPackageManifest[];
  error?: Error;
  getCSVs: (namespace: string) => Promise<IClusterServiceVersion[]>;
  csvs: IClusterServiceVersion[];
  filter: string;
  pushSearchFilter: (filter: string) => RouterAction;
}

export interface IOperatorListState {
  filter: string;
}

class OperatorList extends React.Component<IOperatorListProps, IOperatorListState> {
  public state: IOperatorListState = {
    filter: "",
  };

  public componentDidMount() {
    this.props.checkOLMInstalled();
    this.props.getOperators(this.props.namespace);
    this.props.getCSVs(this.props.namespace);
    this.setState({ filter: this.props.filter });
  }

  public componentDidUpdate(prevProps: IOperatorListProps) {
    if (prevProps.namespace !== this.props.namespace) {
      this.props.getOperators(this.props.namespace);
      this.props.getCSVs(this.props.namespace);
    }
    if (this.props.filter !== prevProps.filter) {
      this.props.getOperators(this.props.namespace);
      this.props.getCSVs(this.props.namespace);
      this.setState({ filter: this.props.filter });
    }
  }

  public render() {
    const { isFetching, pushSearchFilter } = this.props;
    return (
      <div>
        <PageHeader>
          <h1>Operators</h1>
          <SearchFilter
            className="margin-l-big"
            placeholder="search operators..."
            onChange={this.handleFilterQueryChange}
            value={this.state.filter}
            onSubmit={pushSearchFilter}
          />
        </PageHeader>
        <main>
          <MessageAlert level="warning">
            <div>
              Operators integration is under heavy development and currently in alpha state. If you
              find an issue please report it{" "}
              <a target="_blank" href="https://github.com/kubeapps/kubeapps/issues">
                here.
              </a>
            </div>
          </MessageAlert>
          <LoadingWrapper loaded={!isFetching}>{this.renderOperators()}</LoadingWrapper>
        </main>
      </div>
    );
  }

  private renderOperators() {
    const { operators, error, csvs, isOLMInstalled } = this.props;
    const { filter } = this.state;
    if (error && error.constructor === ForbiddenError) {
      return (
        <ErrorSelector
          error={error}
          action="list"
          resource="Operators"
          namespace={this.props.namespace}
        />
      );
    }
    if (!isOLMInstalled) {
      return <OLMNotFound />;
    }
    if (error) {
      return (
        <ErrorSelector
          error={error}
          action="list"
          resource="Operators"
          namespace={this.props.namespace}
        />
      );
    }
    const csvNames = csvs.map(csv => csv.metadata.name);
    const installedOperators: IPackageManifest[] = [];
    const availableOperators: IPackageManifest[] = [];
    const filteredOperators = operators.filter(c =>
      new RegExp(escapeRegExp(filter), "i").test(c.metadata.name),
    );
    if (filteredOperators.length === 0) {
      return <p>No Operator found</p>;
    }
    filteredOperators.forEach(operator => {
      const defaultChannel = operator.status.defaultChannel;
      const channel = operator.status.channels.find(ch => ch.name === defaultChannel);
      if (csvNames.some(csvName => csvName === channel?.currentCSV)) {
        installedOperators.push(operator);
      } else {
        availableOperators.push(operator);
      }
    });
    return (
      <>
        {installedOperators.length > 0 && (
          <>
            <h3>Installed</h3>
            <CardGrid>
              {installedOperators.map(operator => {
                return (
                  <InfoCard
                    key={operator.metadata.name}
                    link={app.operators.view(this.props.namespace, operator.metadata.name)}
                    title={operator.metadata.name}
                    icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
                    info={`v${operator.status.channels[0].currentCSVDesc.version}`}
                    tag1Content={operator.status.channels[0].currentCSVDesc.annotations.categories}
                    tag2Content={operator.status.provider.name}
                  />
                );
              })}
            </CardGrid>
          </>
        )}
        <h3>Available Operators</h3>
        <CardGrid>
          {availableOperators.map(operator => {
            return (
              <InfoCard
                key={operator.metadata.name}
                link={app.operators.view(this.props.namespace, operator.metadata.name)}
                title={operator.metadata.name}
                icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
                info={`v${operator.status.channels[0].currentCSVDesc.version}`}
                tag1Content={operator.status.channels[0].currentCSVDesc.annotations.categories}
                tag2Content={operator.status.provider.name}
              />
            );
          })}
        </CardGrid>
      </>
    );
  }

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };
}

export default OperatorList;
