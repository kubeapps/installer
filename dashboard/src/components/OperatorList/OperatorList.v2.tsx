import actions from "actions";
import CardGrid from "components/Card/CardGrid.v2";
import { CdsIcon } from "components/Clarity/clarity";
import FilterGroup from "components/FilterGroup/FilterGroup";
import Alert from "components/js/Alert";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { RouterAction } from "connected-react-router";
import { flatten, intersection, uniq } from "lodash";
import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { IPackageManifestStatus, IStoreState } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";
import {
  AUTO_PILOT,
  BASIC_INSTALL,
  DEEP_INSIGHTS,
  FULL_LIFECYCLE,
  SEAMLESS_UPGRADES,
} from "../OperatorView/OperatorCapabilityLevel";
import PageHeader from "../PageHeader/PageHeader.v2";
import SearchFilter from "../SearchFilter/SearchFilter.v2";
import OLMNotFound from "./OLMNotFound.v2";
import OperatorItems from "./OperatorItems";
import OperatorNotSupported from "./OperatorsNotSupported.v2";

export interface IOperatorListProps {
  cluster: string;
  namespace: string;
  filter: string;
  pushSearchFilter: (filter: string) => RouterAction;
}

export interface IOperatorListState {
  filter: string;
  categories: string[];
  filterCategories: { [key: string]: boolean };
  filterCapabilities: { [key: string]: boolean };
}

function getDefaultChannel(packageStatus: IPackageManifestStatus) {
  const defaultChannel = packageStatus.defaultChannel;
  const channel = packageStatus.channels.find(ch => ch.name === defaultChannel);
  return channel!;
}

function getCategories(packageStatus: IPackageManifestStatus) {
  const channel = getDefaultChannel(packageStatus);
  return channel.currentCSVDesc.annotations.categories.split(",").map(c => c.trim());
}

function getCapabilities(packageStatus: IPackageManifestStatus) {
  const channel = getDefaultChannel(packageStatus);
  return channel.currentCSVDesc.annotations.capabilities;
}

export default function OperatorList({
  cluster,
  namespace,
  filter,
  pushSearchFilter,
}: IOperatorListProps) {
  const dispatch = useDispatch();
  const [searchFilter, setSearchFilter] = useState(filter);
  const [categoryFilter, setCategoryFilter] = useState([] as string[]);
  const [capabilitiesFilter, setCapabilitiesFilter] = useState([] as string[]);

  useEffect(() => {
    setSearchFilter(filter);
  }, [filter]);

  useEffect(() => {
    dispatch(actions.operators.checkOLMInstalled(namespace));
  }, [dispatch, namespace]);

  const {
    operators: {
      operators,
      isFetching,
      errors: {
        operator: { fetch: error },
      },
      isOLMInstalled,
    },
    config: { kubeappsCluster },
  } = useSelector((state: IStoreState) => state);

  useEffect(() => {
    if (isOLMInstalled) {
      dispatch(actions.operators.getOperators(namespace));
    }
  }, [dispatch, namespace, isOLMInstalled]);

  if (cluster !== kubeappsCluster) {
    return <OperatorNotSupported kubeappsCluster={kubeappsCluster} namespace={namespace} />;
  }

  const allCapabilities = [
    BASIC_INSTALL,
    SEAMLESS_UPGRADES,
    FULL_LIFECYCLE,
    DEEP_INSIGHTS,
    AUTO_PILOT,
  ];
  const allCategories = uniq(
    flatten(operators.map(operator => getCategories(operator.status))),
  ).sort();

  const filteredOperators = operators
    .filter(
      c =>
        capabilitiesFilter.length === 0 || capabilitiesFilter.includes(getCapabilities(c.status)),
    )
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.metadata.name))
    .filter(
      c =>
        categoryFilter.length === 0 || intersection(categoryFilter, getCategories(c.status)).length,
    );

  return (
    <section>
      <PageHeader>
        <Row>
          <h1>Operators</h1>
          <SearchFilter
            key="searchFilter"
            placeholder="search charts..."
            onChange={setSearchFilter}
            value={searchFilter}
            onSubmit={pushSearchFilter}
          />
        </Row>
      </PageHeader>
      <Alert theme="warning">
        <div>
          Operators integration is under heavy development and currently in beta state. If you find
          an issue please report it{" "}
          <a
            target="_blank"
            rel="noopener noreferrer"
            href="https://github.com/kubeapps/kubeapps/issues"
          >
            here.
          </a>
        </div>
      </Alert>
      <LoadingWrapper loaded={!isFetching}>
        {error && <Alert theme="danger">Found en error fetching operators: {error.message}</Alert>}
        {!isOLMInstalled ? (
          <OLMNotFound />
        ) : operators.length === 0 ? (
          <div className="section-not-found">
            <div>
              <CdsIcon shape="bundle" size="64" />
              <h4>The list of Operators is empty</h4>
              <p>
                This may mean that the OLM is still populating the catalog or that it found an
                error. Check the OLM logs for more information.
              </p>
            </div>
          </div>
        ) : (
          <Row>
            <Column span={2}>
              <div className="filters-menu">
                <h5>Filters</h5>
                {allCategories.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Category:</label>
                    <FilterGroup
                      name="category"
                      options={allCategories}
                      onChange={setCategoryFilter}
                    />
                  </div>
                )}
                {allCapabilities.length > 0 && (
                  <div className="filter-section">
                    <label>Application Repository:</label>
                    <FilterGroup
                      name="apprepo"
                      options={allCapabilities}
                      onChange={setCapabilitiesFilter}
                    />
                  </div>
                )}
              </div>
            </Column>
            <Column span={10}>
              <CardGrid>
                <OperatorItems
                  operators={filteredOperators}
                  cluster={cluster}
                  namespace={namespace}
                />
              </CardGrid>
            </Column>
          </Row>
        )}
      </LoadingWrapper>
    </section>
  );
}
