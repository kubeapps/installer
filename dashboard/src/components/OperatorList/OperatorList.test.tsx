import { shallow } from "enzyme";
import * as React from "react";
import { IPackageManifest } from "shared/types";
import itBehavesLike from "../../shared/specs";
import { ErrorSelector } from "../ErrorAlert";
import InfoCard from "../InfoCard";
import OLMNotFound from "./OLMNotFound";
import OperatorList, { IOperatorListProps } from "./OperatorList";

const defaultProps: IOperatorListProps = {
  isFetching: false,
  checkOLMInstalled: jest.fn(),
  isOLMInstalled: false,
  operators: [],
  namespace: "default",
  getOperators: jest.fn(),
};

const sampleOperator = {
  metadata: {
    name: "foo",
  },
  status: {
    provider: {
      name: "kubeapps",
    },
    channels: [
      {
        currentCSVDesc: {
          version: "1.0.0",
          annotations: {
            categories: "security",
          },
        },
      },
    ],
  },
} as IPackageManifest;

itBehavesLike("aLoadingComponent", {
  component: OperatorList,
  props: { ...defaultProps, isFetching: true },
});

it("call the OLM check and render the NotFound message if not found", () => {
  const checkOLMInstalled = jest.fn();
  const wrapper = shallow(<OperatorList {...defaultProps} checkOLMInstalled={checkOLMInstalled} />);
  expect(checkOLMInstalled).toHaveBeenCalled();
  expect(wrapper.find(OLMNotFound)).toExist();
});

it("renders an error if exists", () => {
  const wrapper = shallow(
    <OperatorList {...defaultProps} isOLMInstalled={true} error={new Error("Boom!")} />,
  );
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(
    wrapper
      .find(ErrorSelector)
      .dive()
      .dive()
      .text(),
  ).toMatch("Boom!");
});

it("skips the error if the OLM is not installed", () => {
  const wrapper = shallow(
    <OperatorList
      {...defaultProps}
      isOLMInstalled={false}
      error={new Error("There are no operators!")}
    />,
  );
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(OLMNotFound)).toExist();
});

it("render the operator list if the OLM is installed", () => {
  const wrapper = shallow(
    <OperatorList {...defaultProps} isOLMInstalled={true} operators={[sampleOperator]} />,
  );
  expect(wrapper.find(OLMNotFound)).not.toExist();
  expect(wrapper.find(InfoCard)).toExist();
  expect(wrapper).toMatchSnapshot();
});
