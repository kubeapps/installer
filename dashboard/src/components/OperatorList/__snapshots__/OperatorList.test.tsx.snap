// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <OLMNotFound />
    </LoadingWrapper>
  </main>
</div>
`;

exports[`render the operator list with installed operators 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <h3>
        Installed
      </h3>
      <CardGrid>
        <InfoCard
          icon="api/v1/namespaces/default/operator/foo/logo"
          info="v1.0.0"
          key="foo"
          link="/ns/default/operators/foo"
          tag1Content="security"
          tag2Content="kubeapps"
          title="foo"
        />
      </CardGrid>
      <h3>
        Available Operators
      </h3>
      <CardGrid />
    </LoadingWrapper>
  </main>
</div>
`;

exports[`render the operator list without installed operators 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <h3>
        Available Operators
      </h3>
      <CardGrid>
        <InfoCard
          icon="api/v1/namespaces/default/operator/foo/logo"
          info="v1.0.0"
          key="foo"
          link="/ns/default/operators/foo"
          tag1Content="security"
          tag2Content="kubeapps"
          title="foo"
        />
      </CardGrid>
    </LoadingWrapper>
  </main>
</div>
`;
